Fix MatrixFree + FE_Q_iso_Q1(2)

peterrum opened the pull request.

kronbichler commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -4273,7 +4275,8 @@ namespace internal
       else if (fe_degree > -1 &&
                subface_index >= GeometryInfo<dim>::max_children_per_cell &&
-               shape_info.element_type <= MatrixFreeFunctions::tensor_symmetric)
+               shape_info.element_type <=
+                 MatrixFreeFunctions::tensor_symmetric_no_collocation)
         FEFaceEvaluationImpl<true,
```
Isn't this branch also using collocation derivatives: https://github.com/dealii/dealii/blob/2d6e494f91682cc2f737442cff989365f40858ea/include/deal.II/matrix_free/evaluation_kernels.h#L2533-L2534
Hence, I believe we need to use `tensor_symmetric` here and in the other face function below.

kronbichler commented on include/deal.II/matrix_free/evaluation_kernels.h:
```diff
@@ -5174,7 +5178,7 @@ namespace internal
       Assert(fe_degree > -1, ExcInternalError());
       Assert(fe_eval.get_shape_info().element_type <=
-               MatrixFreeFunctions::tensor_symmetric,
+               MatrixFreeFunctions::tensor_symmetric_no_collocation,
```
Here, I think we also need to play safe and only enable it for the `tensor_symmetric` case.

kronbichler commented on include/deal.II/matrix_free/shape_info.templates.h:
```diff
@@ -930,6 +931,10 @@ namespace internal
             }
+
+
+          if (dynamic_cast<const FE_Q_iso_Q1<dim> *>(&fe))
+            element_type = tensor_symmetric_no_collocation;
```
I would prefer if we could move this above the `if` statement for `check_1d_shapes_collocation`, because we should not even build the matrices for that case.

kronbichler commented:
You would need to adjust the output of `tests/matrix_free/shape_info.cc` because the indices of non-symmetric shape functions get shifted.

peterrum commented:
> You would need to adjust the output of tests/matrix_free/shape_info.cc because the indices of non-symmetric shape functions get shifted.

I am working on that and on an additional test, which performs a sanity test.

peterrum commented:
@kronbichler I have applied the suggested changes, modified the test output and added a new test, which tests the influence of dofs on quadrature point values/gradient (a sanity check I used to debug the problem).
